spec/hash_mergers/answer_spec.rb

```diff
@@ -0,0 +1,105 @@
+
+# frozen_string_literal: true
+
+require 'hash_mergers'
+require 'alces_utils'
+require 'data'
+require 'file_path'
+
+RSpec.describe Metalware::HashMergers::Answer do
+  include AlcesUtils
+
+  let :group { AlcesUtils.mock(self) { mock_group('new_group') } }
+  let :node do
+    AlcesUtils.mock(self) { mock_node('new_node', group.name) }
+  end
+
+  let :identifier { :question_identifier }
+  let :questions do
+    {
+      domain: [{
+        identifier: identifier.to_s,
+        question: 'Ask domain question?',
+        default: 'domain-default',
+      }],
+      group: [{
+        identifier: identifier.to_s,
+        question: 'Ask group question?',
+        default: 'group-default', # Should be ignored
+      }],
+      node: [{
+        identifier: identifier.to_s,
+        question: 'Ask node question?',
+        default: 'node-default', # Should be ignored
+      }],
+      local: []
+    }
+  end
+
+  def answers(namespace)
+    case namespace
+    when Metalware::Namespaces::Domain
+      'domain_answer'
+    when Metalware::Namespaces::Group
+      'group_answer'
+    when Metalware::Namespaces::Node
+      'node_answer'
+    else
+      raise 'unexpected error'
+    end
+  end
+
+  before :each do
+    Metalware::Data.dump Metalware::FilePath.configure_file, questions
+  end
+
+  shared_examples 'run contexts with shared' do |spec_group|
+    context 'when loading domain answers' do
+      subject { alces.domain }
+      include_examples spec_group
+    end
+
+    context 'when loading group answers' do
+      subject { group }
+      include_examples spec_group
+    end
+
+    context 'when loading node answers' do
+      subject { node }
+      include_examples spec_group
+    end
+  end
+
+  context 'without answer files' do
+    shared_examples 'uses the domain default' do
+      it 'uses the domain default' do
+        expect(subject.answer.send(identifier)).to eq('domain-default')
+      end
+    end
+    include_examples 'run contexts with shared', 'uses the domain default'
+  end
+
+  context 'with answer files' do
+    before :each do
+      Metalware::Data.dump(
+        Metalware::FilePath.domain_answers,
+        identifier => answers(alces.domain)
+      )
+      Metalware::Data.dump(
+        Metalware::FilePath.group_answers(group.name),
+        identifier => answers(group)
+      )
+      Metalware::Data.dump(
+        Metalware::FilePath.node_answers(node.name),
+        identifier => answers(node)
+      )
+    end
+
+    shared_examples 'uses the saved answer' do
+      it 'uses the saved answer' do
+        expect(subject.answer.send(identifier)).to eq(answers(subject))
+      end
+    end
+    include_examples 'run contexts with shared', 'uses the saved answer'
```

Comment:
I find the amount of indirection in this file quite confusing, with shared examples being passed other shared examples which they then include in various places, although I understand why you're doing it. Possibly we can improve it a bit though. A couple of suggestions you could try:

- Rather than have the name of a shared example be passed to `include_examples 'run contexts with shared'`, it would be clearer I think to pass a hash of the answers which should be used for each namespace. I.e. this would be a hash of each namespace to `'domain-default'` for `'without answer files'`, and a hash of each namespace to the answers as defined in `answers` for `'with answer files'`. This would also DRY up the repeated `expect(subject.answer.send(identifier)).to eq(...)` in each of the passed shared examples.

- I don't think `'run contexts with shared'` is a very clear shared example group name, as it uses the language of RSpec rather than the language of what we are actually testing, which would be more descriptive. It would also become inaccurate following the above suggested change. Possibly 'uses answers for namespaces' would then be a more descriptive name?

Reply:
This works for now, I have created a card for it:
https://trello.com/c/BnJaqzqE/251-clean-up-answer-spec

Reply:
Thanks :slightly_smiling_face: